enable usb weak host workaround to improve android stability

gwen2018 opened the pull request:
Enable L515 android stability workaround. Current usb buffer settings in device require large number of small usb transactions from device to usb host controller, some hosts with relatively weak host cannot handle this many transactions quickly, results in stability issues, camera disappear from os enumeration or one of the streams dies during stream, etc. This workaround uses larger buffer settings to reduce the number of transaction and improve performance and stability.

Workaround is enabled by default for all android platforms. Requires additional firmware command support in new firmware release.

Tracked on: RS5-8011

maloel commented on src/l500/l500-color.cpp:
```diff
@@ -105,6 +105,16 @@ namespace librealsense
+#ifdef __ANDROID__
+        // option to enable workaround for help weak usb hosts to support L515 devices
+        // with improved performance and stability
+        if(_fw_version >= firmware_version( "1.4.0.66" ) )
```
This will catch FW 1.5.0.0 (the actual released FW currently) which does NOT have the changes required. Please work with FW team to get a number (1.5+) from them that you're sure will work.

maloel replied:
(Same in the depth sensor)

gwen2018 replied:
don't know the actual fw version yet, I put in 1.5.1.0 for now, may need to do another PR to change it.

maloel commented on src/l500/l500-color.cpp:
```diff
@@ -358,6 +368,33 @@ namespace librealsense
+                    command cmdTprocThresholdEp5(ivcam2::TPROC_TRB_THRSLD_SET, 5, 1);
+                    _owner->_hw_monitor->send(cmdTprocThresholdEp5);
+
+                    LOG_INFO("Color usb tproc granularity and TRB threshold updated.");
```
Should be debug -- no need to see it in the Viewer.

gwen2018 replied:
fixed

maloel commented on src/l500/l500-color.cpp:
```diff
@@ -358,6 +368,33 @@ namespace librealsense
+
+            if (usb_perf_enabled)
+            {
+                // TPROC USB Granularity and TRB threshold settings for improved performance and stability
```
Please include the ticket number (RS5-8011), both here and in the depth sensor, so we can go back and read what this is all about

gwen2018 replied:
included ticket number

maloel commented on src/l500/l500-color.cpp:
```diff
@@ -105,6 +105,16 @@ namespace librealsense
+        // with improved performance and stability
+        if(_fw_version >= firmware_version( "1.4.0.66" ) )
+        {
+            color_ep->register_option(RS2_OPTION_ENABLE_WEAK_USB_HOST_WA, std::make_shared<librealsense::float_option>(option_range{ 0, 1, 1, 1 }));
```
Consider using bool_option: would make this a little more readable

maloel replied:
So this defaults to on, regardless of processor/environment? In all Android compilations? That's the way we want it?

gwen2018 replied:
used the float option in case we need to set more level of settings, currently just a single setting.

yes, we would like to turn this on by default on all android devices, because we found other platform like pixel may have similar issue, just in much less severe degree.

maloel commented on src/sensor.cpp:
```diff
@@ -471,6 +471,19 @@ namespace librealsense
             throw wrong_api_call_sequence_exception("start_streaming(...) failed. UVC device was not opened!");
 
+        // RS2_OPTION_AUTO_EXPOSURE_PRIORITY will be disabled by default in firmware
+        // temporarily disable here to facilitate android stability tests
```
To facilitate stability tests? Is this not meant to get into the code?

gwen2018 replied:
removed.

gwen2018 commented:
@maloel thanks for review feedback. hope the changes addressed concerns.

maloel commented:
Hi @gwen2018 
The changes are fine. I'm a little hesitant about the FW version... can we delay the merge of this until they give a proper version? Or do you need it in right away?

gwen2018 commented:
> Hi @gwen2018
> The changes are fine. I'm a little hesitant about the FW version... can we delay the merge of this until they give a proper version? Or do you need it in right away?

Thanks. I will check with firmware team.

ev-mp commented on include/librealsense2/h/rs_option.h:
```diff
@@ -96,6 +96,7 @@ extern "C" {
         RS2_OPTION_THERMAL_COMPENSATION, /**< Depth Thermal Compensation for selected D400 SKUs */
         RS2_OPTION_TRIGGER_CAMERA_ACCURACY_HEALTH, /**< Enable depth & color frame sync with periodic calibration for proper alignment */
         RS2_OPTION_RESET_CAMERA_ACCURACY_HEALTH,
+        RS2_OPTION_ENABLE_WEAK_USB_HOST_WA, /**< Enable workaround for weak USB hosts to improve performance and stability */
```
Pls rename and refactor the option to be rs2_option_trb_size so that it will read/set the TRB size (units TBD) - it is not platform specfic and therefore may be utilized in additional scenarios.
You can set an enum for the supported RGB sizes in order to enforce strict (and validated) values

ev-mp commented on src/l500/l500-depth.cpp:
```diff
@@ -374,6 +383,44 @@ namespace librealsense
+
+            auto usb_perf_enabled = get_option(RS2_OPTION_ENABLE_WEAK_USB_HOST_WA).query();
+
+            if (usb_perf_enabled)
```
It should be encapsulated into opt->set

gwen2018 commented:
@ev-mp thanks for the review. refactored the option, included your suggestion, name is host performance mode, because it will be a generic option for future use as well, this time, only include the TRB settings.

at launch, on android platforms, low performance mode is turned on and larger TRB settings, on other platforms (windows, linux, etc.), device default setting is used at launch. user has the option to change the option later after viewer launch.

The option is only for L515, the conditional compilation is minimized (only the launch performance mode).

also added settings for the 4th endpoint (confidence stream). now the settings are complete.

ev-mp commented on src/l500/l500-depth.cpp:
```diff
@@ -374,6 +392,67 @@ namespace librealsense
 
     void l500_depth_sensor::start(frame_callback_ptr callback)
     {
+        if (supports_option(RS2_OPTION_HOST_PERFORMANCE))
```
What about encapsulating the below code into opt->set(..) ?
The TRB settings are customized per sensor and can be passed into the option at ctor level (where sensor->register_option( occrs), so this could can become:
if supports(TRB)
  trb_opt-set(TRB_MOD)

gwen2018 replied:
thanks for the suggestion, looked into it, will implement tomorrow. if need to make the build, please go ahead and take what current implementation. can do another PR for this part of the change once it's tested.

ev-mp replied:
Thanks, we'll merge the current implementation so that refactoring will be moved to the follow-on task.
@gwen2018 , please open a new PR for the remaining work.

ev-mp commented on include/librealsense2/h/rs_option.h:
```diff
@@ -96,6 +96,7 @@ extern "C" {
         RS2_OPTION_THERMAL_COMPENSATION, /**< Depth Thermal Compensation for selected D400 SKUs */
         RS2_OPTION_TRIGGER_CAMERA_ACCURACY_HEALTH, /**< Enable depth & color frame sync with periodic calibration for proper alignment */
         RS2_OPTION_RESET_CAMERA_ACCURACY_HEALTH,
+        RS2_OPTION_HOST_PERFORMANCE, /**< Set host performance mode to optimize device settings so host can keep up with workload, for example, USB transaction granularity, setting option to low performance host leads to larger USB transaction size and reduced number of transactions which improves performance and stability if host is relatively weak as compared to workload */
```
Make sure the option and enum are propagated into python (+pybackend), C#, Java and Matlab.

gwen2018 replied:
added to python and java for now, but not c# and matlab yet.

ev-mp commented on include/librealsense2/h/rs_option.h:
```diff
@@ -175,6 +176,16 @@ extern "C" {
+        RS2_HOST_PERF_HIGH = 2,     /**< high performance host mode, if host is strong as compared to the work and can handle workload without delay, this option sets smaller USB transactions granularity and as result larger number of transactions and workload on host, but reduces chance in device frame drops */
+        RS2_HOST_PERF_COUNT         /**< Number of enumeration values. Not a valid input: intended to be used in for-loops. */
+    } rs2_host_perf_mode;
+    const char* rs2_host_perf_mode_to_string( rs2_host_perf_mode perf );
```
This should be added into the public API - realsense.def and then added into the wrappers (above)

gwen2018 replied:
added.

ev-mp reviewed (approved):
Looks good, thank you for the in-depth analysis!
